bpo-40529: rlcompleter with case insensitive

kasulamadhusudhan opened the pull request:
Added an option to the user to make completions with case insensitive / sensitive.
rlcompleter.set_case_insensitive(True) # to enable case insensitive
Added documentation accordingly.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-40529](https://bugs.python.org/issue40529) -->
https://bugs.python.org/issue40529
<!-- /issue-number -->

kasulamadhusudhan committed:
bpo-40529: rlcompleter with case insensitive

Added an option to the user to make completions with case insensitive / sensitive.
rlcompleter.set_case_insensitive(True) # to enable case insensitive
Added documentation accordingly.

* skip news

remilapeyre commented on Lib/rlcompleter.py:
```diff
@@ -185,6 +189,12 @@ def attr_matches(self, text):
+def set_case_insensitive(option):
+    import re
+    global _re_ignorecase_flags
+    _re_ignorecase_flags = option and re.IGNORECASE or 0
```
```suggestion
    _re_ignorecase_flags = re.IGNORECASE if option else 0
```

kasulamadhusudhan replied:
Thanks for reviewing. I have modified the line as per your suggestion. And also added Documentation and unit test case.

remilapeyre reviewed (changes requested):
Hi @kasulamadhusudhan, thanks for your contribution! You will have to sign the CLA to get this change accepted if you did not do it yet.

This change will also need to be documented at `Doc/library/rlcompleter.rst` and have some tests before it can be accepted.

kasulamadhusudhan committed:
bpo-40529: rlcompleter with case insensitive

Incorporated review comments @ Lib/rlcompleter.py
Added documentation @ Doc/library/rlcompleter.rst
Added unit testcase test_case_insentive_matches() @ Lib/test/test_rlcompleter.py

* skip news

remilapeyre commented on Lib/rlcompleter.py:
```diff
@@ -185,6 +191,12 @@ def attr_matches(self, text):
         return matches
 
+_re_ignorecase_flags = 0
+def set_case_insensitive(option):
```
I find it weird that the way to be case sensitive is to have a double negation:

```python
set_case_insensitive(False)
```

can we change it to:

```python
set_case_sensitive(True)
```

?

kasulamadhusudhan replied:
I like your comment. If you see the manual page of readline() it is also named to have double negation: 
`completion-ignore-case (Off)`
`       If set to On, readline performs filename matching and completion in a case-insensitive fashion.`

Instead, I would like to name it more similar to readline() as set_ignore_case().

remilapeyre replied:
`set_ignore_case` seems like a better name yes.

kasulamadhusudhan replied:
Thanks. Modified the function name as suggested.

kasulamadhusudhan committed:
bpo-40529: rlcompleter with case insensitive

Changed func name set_case_insensitive() -> set_ignore_case()
Modified Docs and Tests accordingly

* skip news
